pulsar-client/src/main/java/org/apache/pulsar/client/api/Consumer.java

```diff
@@ -278,4 +277,19 @@
      * @return a future to track the completion of the seek operation
      */
     CompletableFuture<Void> seekAsync(MessageId messageId);
+
+    /**
+     * Gets last message id of Topic.
+     *
+     * @return the last message id
+     */
+    MessageId getLastMessageId() throws PulsarClientException;
```

Comment:
Rather than exposing this method, I would prefer to have a simpler way to check it, for a couple of reasons: 
 1. Using and comparing message ids can be tricky (was that < or > ? :) )
```java
MessageId lastMessageId = consumer.getLastMessageId();
MessageId currentMessageId = null;
while (currentMessageId == null || currentMessageId.compareTo(lastMessageId) <= 0) {
     Message msg = consumer.receive();
     ///// 
     consumer.acknowledge(msg);
}
```
 2. Most typical code would be like the above snippet and do a tight loop asking the broker for the last message id, without any throttling or flow-control

Rather, I would prefer something like: 

```java
while (consumer.hasMessageAvaiable()) {
    Message msg = consumer.receive();
    // Do something
    consumer.acknowledge(msg);
}

// Exits loops when the consumer has caught up with the writer
```

Internally, the client library can ask the broker for the last message id, and cache it. For example if the `lastMessageId=123:12` but the consumer is currently receiving `110:5`, there's no need to keep asking the broker.

Reply:
hmm `consumer.hasMessageAvaiable()` - I am not sure it is a good idea. However it introduces non-deterministic behaviour to users, where user will have no idea where to stop. but I agree rate-limiting is also important, however should we consider rate-limiting at broker side for this message, rather than do the rate-limiting at client side. In that way, it is a probably a better solution for long term. thoughts?

Reply:
In the above example, rate limiting in broker will slow down the consumption

Reply:
Thanks for all the comments. 
If needed, will add a position var to track the last position in`incomingMessages` to achieve `consumer.hasMessageAvaiable()`.